File: typescript-agent/src/utils/signature.ts
import crypto from "crypto";

export function verifySignature(
  raw: Buffer,
  signature: string,
  timestamp: string,
  secret: string,
) {
  const age = Math.abs(Date.now() - new Date(timestamp).getTime()) / 1000;
  if (age > 300) return false;

  const expected = crypto
    .createHmac("sha256", secret)
    .update(`${timestamp}.${raw.toString()}`)
    .digest("hex");

  try {
    return crypto.timingSafeEqual(
      Buffer.from(`sha256=${expected}`),
      Buffer.from(signature),
    );
  } catch {
    return false;
  }
}

File: python-agent/app/utils/signature.py
import hashlib
import hmac
from datetime import datetime, timezone


def verify_signature(raw: bytes, signature: str, timestamp: str, secret: str) -> bool:
    if not signature or not timestamp or not secret:
        return False

    try:
        ts = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        now = datetime.now(timezone.utc)
        age_sec = abs((now - ts).total_seconds())
    except Exception:
        return False

    if age_sec > 300:
        return False

    try:
        body_text = raw.decode("utf-8")
    except Exception:
        return False

    expected = hmac.new(
        secret.encode("utf-8"),
        f"{timestamp}.{body_text}".encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return hmac.compare_digest(f"sha256={expected}", signature)
